api/service/explorer/storage.go

```diff
@@ -86,71 +86,95 @@ func (storage *Storage) Dump(block *types.Block, height uint64) {
 	// Store txs
 	for _, tx := range block.Transactions() {
 		explorerTransaction := GetTransaction(tx, block)
-		storage.UpdateTxAddress(explorerTransaction, tx)
+		storage.UpdateTxAddressStorage(explorerTransaction, tx)
 	}
 	// Store staking txns
 	for _, tx := range block.StakingTransactions() {
 		explorerTransaction := GetStakingTransaction(tx, block)
-		storage.UpdateStakingTxAddress(explorerTransaction, tx)
+		storage.UpdateStakingTxAddressStorage(explorerTransaction, tx)
 	}
 }
 
-// UpdateTxAddress ...
-func (storage *Storage) UpdateTxAddress(explorerTransaction *Transaction, tx *types.Transaction) {
-	explorerTransaction.Type = Received
-	if explorerTransaction.To != "" {
-		storage.UpdateTxAddressStorage(explorerTransaction.To, explorerTransaction, tx)
-	}
-	explorerTransaction.Type = Sent
-	storage.UpdateTxAddressStorage(explorerTransaction.From, explorerTransaction, tx)
-}
-
-// UpdateStakingTxAddress ...
-func (storage *Storage) UpdateStakingTxAddress(explorerTransaction *StakingTransaction, tx *staking.StakingTransaction) {
-	explorerTransaction.Type = Received
-	if explorerTransaction.To != "" {
-		storage.UpdateStakingTxAddressStorage(explorerTransaction.To, explorerTransaction, tx)
-	}
+// UpdateTxAddressStorage stores Transaction to explorer storage ldb
+func (storage *Storage) UpdateTxAddressStorage(explorerTransaction *Transaction, tx *types.Transaction) {
+	// sent transaction
 	explorerTransaction.Type = Sent
-	storage.UpdateStakingTxAddressStorage(explorerTransaction.From, explorerTransaction, tx)
-}
-
-// UpdateTxAddressStorage updates specific addr tx Address.
-func (storage *Storage) UpdateTxAddressStorage(addr string, explorerTransaction *Transaction, tx *types.Transaction) {
-	var address Address
-	key := GetAddressKey(addr)
-	if data, err := storage.db.Get([]byte(key), nil); err == nil {
-		if err = rlp.DecodeBytes(data, &address); err != nil {
-			utils.Logger().Error().Err(err).Msg("Failed due to error")
+	f := func(explorerTransaction *Transaction) {
+		var address Address
+		var key string
+		var addr []byte
+		var data []byte
+		var err error
+
+		key = GetAddressKey(explorerTransaction.From)
+		addr = []byte(key)
+
+		if data, err = storage.db.Get(addr, nil); err == nil {
```

Comment:
need to ensure atomicity on read-update-read db operation.

Reply:
would go channel process functions linearly? in the channel queue?

Reply:
go goroutine is just a thread. it could happen that multiple threads read/writing to the same data. I would say the initial linear process is better in terms of stability.